drivers/cc110x/cc110x_netdev.c

```diff
@@ -0,0 +1,610 @@
+    mutex_init(&dev->isr_signal);
+    mutex_lock(&dev->isr_signal);
+
+    /* Make sure the crystal is stable and the chip ready */
+    if (cc110x_power_on(dev)) {
+        DEBUG("[cc110x] netdev_driver_t::init(): Failed to pull CS pin low\n");
+        return -EIO;
+    }
+
+    if (cc110x_acquire(dev) != SPI_OK) {
+        DEBUG("[cc110x] netdev_driver_t::init(): Failed to setup/acquire SPI "
+              "interface\n");
+        return -EIO;
+    }
+
+    /* Performing a reset of the transceiver to get it in a known state */
+    cc110x_cmd(dev, CC110X_STROBE_RESET);
+    cc110x_release(dev);
+
+    /* Again, make sure the crystal is stable and the chip ready */
+    if (cc110x_power_on(dev)) {
+        DEBUG("[cc110x] netdev_driver_t::init(): Failed to pull CS pin low "
+              "after reset\n");
+        return -EIO;
+    }
+
+    if (cc110x_acquire(dev) != SPI_OK) {
+        DEBUG("[cc110x] netdev_driver_t::init(): Failed to setup/acquire SPI "
+              "interface after reset\n");
+        return -EIO;
+    }
+
+    if (check_device(dev)) {
+        DEBUG("[cc110x] netdev_driver_t::init(): Device check failed\n");
+        cc110x_release(dev);
+        return -ENOTSUP;
+    }
+
+    /* Upload the main configuration */
+    cc110x_burst_write(dev, CC110X_CONF_START, cc110x_conf, CC110X_CONF_SIZE);
+    /* Set TX power to match uploaded configuration */
+    dev->tx_power = CC110X_TX_POWER_0_DBM;
+
+    /* Upload the poorly documented magic numbers obtained via SmartRF Studio */
+    cc110x_burst_write(dev, CC110X_REG_TEST2, cc110x_magic_registers,
+                       sizeof(cc110x_magic_registers));
+
+    /* Setup the selected PA_TABLE */
+    cc110x_burst_write(dev, CC110X_MULTIREG_PATABLE,
+                       dev->params.patable->data, CC110X_PATABLE_LEN);
+
+    /* Verify main config, magic numbers and PA_TABLE correctly uploaded */
+    if (check_config(dev)) {
+        cc110x_release(dev);
+        return -EIO;
+    }
+
+    /* Verify that pins GDO2 and GDO0 are correctly connected */
+    if (check_gdo_pins(dev)) {
+        cc110x_release(dev);
+        return -EIO;
+    }
+
+    /* Setup the layer 2 address, but do not accept CC110X_L2ADDR_AUTO (which
+     * has the value 0x00 and is used for broadcast)
+     */
+    dev->addr = dev->params.l2addr;
+    while (dev->addr == CC110X_L2ADDR_AUTO) {
+        luid_get(&dev->addr, 1);
+    }
+    cc110x_write(dev, CC110X_REG_ADDR, dev->addr);
+
+    /* Setup interrupt on GDO0  */
+    if (gpio_init_int(dev->params.gdo0, GPIO_IN, GPIO_BOTH,
+                      cc110x_on_gdo, dev)) {
+        cc110x_release(dev);
+        DEBUG("[cc110x] netdev_driver_t::init(): Failed to setup interrupt on "
+              "GDO0 pin\n");
+        return -EIO;
+    }
+
+    /* Setup interrupt on GDO2 */
+    if (gpio_init_int(dev->params.gdo2, GPIO_IN, GPIO_BOTH,
+                      cc110x_on_gdo, dev)) {
+        gpio_irq_disable(dev->params.gdo0);
```

Comment:
`gpio_init_init()` returns `0` on success. So the "then-case" is taken on failure.

The idea is to clean up after a failure to bring back the system as closely as possible to the state before `cc110x_init()` was called. When the initialization of GDO2 fails, the interrupt on GDO0 has already been set up. Therefore, the code disables it again so that no left-over interrupt handlers are registered when init fails.